### PR TITLE
[SPARK-32194][PYTHON] Use proper exception classes instead of plain Exception

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to use a proper built-in exceptions instead of the plain `Exception` in Python.

While I am here, I fixed another minor issue at `DataFrams.schema` together:

```diff
- except AttributeError as e:
-     raise Exception(
-         "Unable to parse datatype from schema. %s" % e)            
+ except Exception as e:
+     raise ValueError(
+         "Unable to parse datatype from schema. %s" % e) from e
```

Now it catches all exceptions during schema parsing, chains the exception with `ValueError`. Previously it only caught `AttributeError` that does not catch all cases.

### Why are the changes needed?

For users to expect the proper exceptions.

### Does this PR introduce _any_ user-facing change?

Yeah, the exception classes became different but should be compatible because previous exception was plain `Exception` which other exceptions inherit.

### How was this patch tested?

Existing unittests should cover,

Closes #31238

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
I tested it locally but let me wait for the GA build results.

File: python/pyspark/context.py [hunk 1]
```diff
@@ -188,9 +188,9 @@ def _do_init(self, master, appName, sparkHome, pyFiles, environment, batchSize,
 
         # Check that we have at least the required parameters
         if not self._conf.contains("spark.master"):
-            raise Exception("A master URL must be set in your configuration")
+            raise TypeError("A master URL must be set in your configuration")
         if not self._conf.contains("spark.app.name"):
-            raise Exception("An application name must be set in your configuration")
+            raise TypeError("An application name must be set in your configuration")
```
Comment (viirya) [inline python/pyspark/context.py:193]: Seems like `ValueError` or `RuntimeError`?

**Review by viirya — APPROVED** [review]
One place I am not sure `TypeError` is a good choise, otherwise lgtm.

File: python/pyspark/context.py [hunk 2]
```diff
@@ -188,9 +188,9 @@ def _do_init(self, master, appName, sparkHome, pyFiles, environment, batchSize,
 
         # Check that we have at least the required parameters
         if not self._conf.contains("spark.master"):
-            raise Exception("A master URL must be set in your configuration")
+            raise TypeError("A master URL must be set in your configuration")
```
Comment (HyukjinKwon) [inline python/pyspark/context.py:191]:
```suggestion
            raise ValueError("A master URL must be set in your configuration")
```

File: python/pyspark/context.py (hunk 1 above)
Comment (HyukjinKwon) [inline python/pyspark/context.py:193]:
```suggestion
            raise ValueError("An application name must be set in your configuration")
```

File: python/pyspark/context.py (hunk 2 above)
Comment (HyukjinKwon) [inline python/pyspark/context.py:191]:
```suggestion
            raise ValueError("A master URL must be set in your configuration")
```

**Comment by HyukjinKwon** [comment]
Merged to master!

**Pull request closed without merging**